### ISSUE TITLE
Intermittent test failures after `Publish AS::Executor and AS::Reloader APIs`

### ISSUE THREAD

**Issue by ojab** [rails/rails#24004]
Looks like d3c9d80 (according to my bisect) leads to intermittent test failures in my [in-house] app.
Test [rspec + factory_girl] that looks in a nutshell like this:
```ruby
RSpec.feature 'Foos' do
  let!(:foo) { create(:foo ) }
  context 'My foos' do
    it do
      p Foo.all
      p foo
      p foo.new_record?
      p Foo.all
    end
  end
end
```
sometimes produces output:
```
#<ActiveRecord::Relation []>
#<Foo id: 9, name: …>
false
#<ActiveRecord::Relation []>
```

Unfortunately I'm unable to make reduced testcase, maybe you can point me how to debug this further.
cc @matthewd

**Comment by ojab** [comment]
I've came up with [kind-of reduced testcase](https://github.com/ojab/24004) (note that it's using poltergeist, so phantomjs should be installed), `bundle exec rspec` produces
```
$ bundle exec rspec

Foos
  whatever again
    testing
      example at ./spec/features/foos_spec.rb:10
  whatever
    testintg
#<ActiveRecord::Relation []>
#<Foo id: 1, created_at: "2016-03-02 19:33:42", updated_at: "2016-03-02 19:33:42">
false
#<ActiveRecord::Relation []>
      example at ./spec/features/foos_spec.rb:19

Finished in 0.8988 seconds (files took 2.07 seconds to load)
2 examples, 0 failures
```
output.

It seems that database_cleaner/capybara are also involved.
AFAIU it happens if:
1. `database_strategy = :truncation` is set for `js: true` tests
2. `js: false` testcase with reproducible issue is called after `js: true` one
3. Both testcases `visit` some location

**Comment by jeremy** [comment]
@ojab Could you try with cf075d9f55047281a5912e1b28cc4cb4bfcfdbf0 ?

**Comment by ojab** [comment]
Still the same, issue exists.

**Comment by arthurnn** [comment]
@ojab I am trying to run your test app.. However routes file is empty, and `visit /` is failing https://github.com/ojab/24004/blob/master/spec/features/foos_spec.rb#L6

**Comment by ojab** [comment]
Works fine here. Do you have phantomjs installed? What exact error do you have?

**Comment by jeremy** [comment]
Fixed by #24610
